pynucastro/templates/amrexastro-cxx-microphysics/table_rates.H.template

```diff
@@ -147,167 +126,211 @@ int vector_index_lu(const int vlen, const V& vector, const Real fvar)
 
-    // Perform bilinear interpolation within the interval [xlo, xhi]
-    // where the function values at the endpoints are defined by:
-    // flo = f(xlo)
-    // fhi = f(xhi)
-    // Returns f(x), the values flo and fhi interpolated at x
-    // f(x) = flo if x <= xlo
-    // f(x) = fhi if x >= xhi
-
-    Real f;
-    if (x <= xlo) {
-        f = flo;
-    } else if (x >= xhi ) {
-        f = fhi;
-    } else {
-        f = (flo * (xhi - x) + fhi * (x - xlo) ) / (xhi - xlo);
-    }
+    int irhoy = vector_index_lu(table_meta.nrhoy, log_rho_table, log_rhoy);
+    int jtemp = vector_index_lu(table_meta.ntemp, log_temp_table, log_temp);
 
-    return f;
-}
+    const Real dlog_rhoy = log_rho_table(irhoy+1) - log_rho_table(irhoy);
+    const Real dlog_temp = log_temp_table(jtemp+1) - log_temp_table(jtemp);
 
-AMREX_INLINE AMREX_GPU_HOST_DEVICE
-Real
-bl_extrap(const Real xlo, const Real xhi, const Real flo, const Real fhi, const Real x)
-{
+    Real fij = data(jtemp, irhoy, component);
+    Real fip1j = data(jtemp, irhoy+1, component);
+    Real fijp1 = data(jtemp+1, irhoy, component);
+    Real fip1jp1 = data(jtemp+1, irhoy+1, component);
 
-    // Perform bilinear interpolation within the interval [xlo, xhi]
-    // where the function values at the endpoints are defined by:
-    // flo = f(xlo)
-    // fhi = f(xhi)
-    // Returns f(x), the values flo and fhi interpolated at x
-    // If x <= xlo or x >= xhi, f(x) is extrapolated at x
+    Real E =  fij;
+    Real C = (fijp1 - fij) / dlog_rhoy;
+    Real B = (fip1j - fij) / dlog_temp;
+    Real A = (fip1jp1 - B * dlog_rhoy - C * dlog_temp + E)/ (dlog_rhoy * dlog_temp);
 
-    Real f = (flo * (xhi - x) + fhi * (x - xlo) ) / (xhi - xlo);
+    Real r =  A * (log_rhoy - log_rho_table(irhoy)) * (log_temp - log_temp_table(jtemp)) +
+              B * (log_rhoy - log_rho_table(irhoy)) +
+              C * (log_temp - log_temp_table(jtemp)) +
+              E;
 
-    return f;
+    return r;
 }
 
 
-template <typename R, typename T, typename D>
+template<typename R, typename T, typename D>
 AMREX_INLINE AMREX_GPU_HOST_DEVICE
-void
-get_entries(const table_t& table_meta,
-            const R& rhoy_table, const T& temp_table, const D& data,
-            const Real rhoy, const Real temp, Array1D<Real, 1, num_vars+1>& entries)
+Real
+evaluate_dr_dtemp(const table_t& table_meta, const R& log_rho_table, const T& log_temp_table, const D& data,
+                  const Real rhoy, const Real temp)
 {
+    //Now we compute dlogr_dlogT
+
+    const Real log_rhoy = std::log10(rhoy);
+    const Real log_temp = std::log10(temp);
+
+    int irhoy = vector_index_lu(table_meta.nrhoy, log_rho_table, log_rhoy);
+    int jtemp = vector_index_lu(table_meta.ntemp, log_temp_table, log_temp);
+
+    const Real dlog_rhoy = log_rho_table(irhoy+1) - log_rho_table(irhoy);
+    const Real dlog_temp = log_temp_table(jtemp+1) - log_temp_table(jtemp);
+
+    Real t_jm1 = log_temp_table(jtemp-1);
```

Comment:
This is an out-of-bounds access if we're at the lower edge of the temperature table (ditto for `fijm1` and `fip1jm1`)

Reply:
Same issue with `t_jp2`, `fijp2`, and `fip1jp2` if we're at the upper edge